src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java

```diff
@@ -61,4 +69,64 @@ public void hideDetails(Submission submission) {
             }
         }
     }
+
+    /**
+     * Maps abstract Submission type to concrete submission and assigns result.
+     * @param result result that will be assigned to concrete submission
+     * @param concreteSubmission concrete submission that will be mapped from abstract submission
+     * @param <T> concrete submission type
+     * @return concrete submission of type T
+     */
+    public <T extends Submission> T mapAbstractToConcreteSubmission(Result result, T concreteSubmission) {
+        Submission submission = result.getSubmission();
+        result.setSubmission(null);
+        if (concreteSubmission instanceof TextSubmission) {
+            concreteSubmission.setLanguage(submission.getLanguage());
+        }
+        concreteSubmission.setResult(result);
+        concreteSubmission.setParticipation(submission.getParticipation());
+        concreteSubmission.setId(submission.getId());
+        concreteSubmission.setSubmissionDate(submission.getSubmissionDate());
+        return concreteSubmission;
+    }
+
+    /**
+     * Creates a new Result object, assigns it to the given submission and stores the changes to the database. Note, that this method is also called for example submissions which
+     * do not have a participation. Therefore, we check if the given submission has a participation and only then update the participation with the new result.
+     *
+     * @param submission the submission for which a new result should be created
+     * @param submissionRepository concrete submission repository
+     * @return the newly created result
+     */
+    public Result setNewResult(Submission submission, GenericSubmissionRepository submissionRepository) {
+        Result result = new Result();
+        result.setSubmission(submission);
+        submission.setResult(result);
+        if (submission.getParticipation() != null) {
+            submission.getParticipation().addResult(result);
+        }
+        resultRepository.save(result);
+        submissionRepository.save(submission);
+        return result;
+    }
+
+    /**
+     * Hides the result details for given submission
+     * @param submission that we want to hide details for
+     * @param exercise to which the submission belongs to
+     * @param <T> concrete submission type
+     * @return submission with result details hidden
+     */
+    public <T extends Submission> T hideResultDetails(T submission, Exercise exercise) {
```

Comment:
I would actually suggest to make the whole class generic with `<T extends Submission>` since we might need more methods like this in the future.

Reply:
Yes, good point.